src/TranslatorSettingsDialog.cpp

```diff
@@ -139,3 +148,73 @@ void TranslatorSettingsDialog::updateModelActions()
     ui_->actionDeleteModel->setEnabled(containsDeletableModel);
     ui_->deleteModelButton->setEnabled(containsDeletableModel);
 }
+
+void TranslatorSettingsDialog::updateRepoActions()
+{
+    bool containsDeletableRepo = false;
+
+    for (auto&& index : ui_->repoTable->selectionModel()->selectedIndexes()) {
+        if (index.row() !=0) { // The first repo is the Bergmaot repo, never deleted
+            containsDeletableRepo = true;
+            break;
+        }
+    }
+
+    ui_->actionDeleteRepo->setEnabled(containsDeletableRepo);
+    ui_->deleteRepo->setEnabled(containsDeletableRepo);
+}
+
+void TranslatorSettingsDialog::on_importRepo_clicked()
+{
+    // Dialog button with two text boxes, because QInputDialog only has one : (
+    QDialog dialog(this);
+    dialog.setWindowTitle("Add new model");
+    // Use a layout allowing to have a label next to each field
+    QFormLayout form(&dialog);
+
+    // Add some text above the fields
+    form.addRow(new QLabel(tr(
+         "Enter the name and URL to the external repository's models.json.\nThe new models will appear in the language selection drop-down.")));
+
+    // Add the lineEdits with their respective labels
+    QLineEdit name(&dialog);
+    name.setToolTip("Enter the name of the new repository here.");
+    QString nameLbl("Name:");
+    form.addRow(nameLbl, &name);
+
+    QLineEdit url(&dialog);
+    url.setToolTip("Enter url to translateLocally like json repository. Eg https://example.com/models.json");
+    QString urlLbl("URL:");
+    form.addRow(urlLbl, &url);
+
+    // Add some standard buttons (Cancel/Ok) at the bottom of the dialog
+    QDialogButtonBox buttonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel,
+                               Qt::Horizontal, &dialog);
+    form.addRow(&buttonBox);
+    QObject::connect(&buttonBox, SIGNAL(accepted()), &dialog, SLOT(accept()));
+    QObject::connect(&buttonBox, SIGNAL(rejected()), &dialog, SLOT(reject()));
+
+    // Show the dialog
+    if (dialog.exec() == QDialog::Accepted) {
+        if (!url.text().isEmpty()) {
+            QStringList new_entry({name.text(), url.text()});
+            modelManager_->getRepoManager()->insert(new_entry);
+        }
+    }
+}
+
+
+void TranslatorSettingsDialog::on_deleteRepo_clicked()
+{
+    auto selection = ui_->repoTable->selectionModel()->selectedRows(0);
+
+    if (QMessageBox::question(this,
+            tr("Remove repo"),
+            tr("Are you sure you want to remove %n selected repo(s)?", "", selection.size())
+        ) != QMessageBox::Yes)
+        return;
+
+    for (auto &&index : selection)
+        modelManager_->getRepoManager()->remove(index);
```

Comment:
Will this work? The indexes are invalidated since the rows might shift after you remove a row.

Reply:
I thought the selection doesn't change until this is executed. I will test removing non-consecutive models (I have tried removing a bunch of entries from the middle of the pack, but always consecutive ones....)

Reply:
(Also just because it works doesn't mean i don't do use-after-free)

Reply:
Will fix, this is wrong. Sorry.